File: 199.BinaryTreeRightSideView.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    const stack = [root];
    let count = 1;
    let current = 0;
    const result = [];
    while(stack.length) {
        const node = stack.shift();
        current++;
        if (node) {
            if (node.left) {
                stack.push(node.left);
            }
            if (node.right) {
                stack.push(node.right);
            }
            if (count === current) {
                result.push(node.val);
                current = 0;
                count = stack.length;
            }
        }
    }
    return result;
};

File: Golang/Tree/199.binary-tree-right-side-view.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rightSideView(root *TreeNode) []int {
	ret := []int{}
	if root != nil {
		stack := []*TreeNode{root}
		for len(stack) > 0 {
			layer := []*TreeNode{}
			for i := 0; i < len(stack); i++ {
				node := stack[i]
				if node.Left != nil {
					layer = append(layer, node.Left)
				}
				if node.Right != nil {
					layer = append(layer, node.Right)
				}
			}
			ret = append(ret, stack[len(stack)-1].Val)
			stack = layer
		}
	}
	return ret
}
